File: desktop/test/forks.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'

import {
  MAX_FORK_DEPTH,
  buildCwdTree,
  buildTree,
  familyRows,
  forksByParent,
  idsOf,
  isNestedFork,
} from '../src/renderer/components/grouping.ts'
import type { Session, SessionGroup } from '../src/shared/models.ts'

const HELIOS: SessionGroup = { key: 'g_helios', name: 'helios', position: 0 }

function root(id: string, cwd = `/x/${id}`, group?: SessionGroup): Session {
  return {
    session_id: id,
    source: 'claude',
    cwd,
    project: id,
    status: 'idle',
    pinned: false,
    sort_order: 0,
    created_at: '2026-09-18T00:00:00Z',
    supports_prompt_queue: false,
    group_path: group ? [group] : undefined,
    group_key: group?.key,
  }
}

function fork(id: string, parent: string, forkedAt: string, cwd = `/x/${id}`, group?: SessionGroup): Session {
  return { ...root(id, cwd, group), forked_from: parent, forked_at: forkedAt }
}

const never = (): boolean => false
const always = (): boolean => true

test('a fork whose parent is in the list nests; one whose parent is missing does not', () => {
  const sessions = [root('a'), fork('b', 'a', '1'), fork('orphan', 'gone', '2')]
  const present = idsOf(sessions)

  assert.equal(isNestedFork(sessions[1]!, present), true)
  assert.equal(isNestedFork(sessions[2]!, present), false, 'an orphan draws as a root')
  assert.equal(isNestedFork(sessions[0]!, present), false)
})

test('siblings are ordered by when the branch was taken, not by arrival', () => {
  const forks = forksByParent([fork('late', 'a', '2026-09-18T12:00:00Z'), fork('early', 'a', '2026-09-18T09:00:00Z')])
  assert.deepEqual(
    forks.get('a')?.map((s) => s.session_id),
    ['early', 'late'],
  )
})

test('a family is drawn depth-first, parent then its own forks', () => {
  const sessions = [
    root('r'),
    fork('a', 'r', '1'),
    fork('deep', 'a', '2'),
    fork('b', 'r', '3'),
  ]
  const rows = familyRows(sessions[0]!, forksByParent(sessions), never)

  assert.deepEqual(
    rows.map((row) => [row.session.session_id, row.depth]),
    [
      ['r', 0],
      ['a', 1],
      ['deep', 2],
      ['b', 1],
    ],
  )
})

test('the trunk says which ancestor lines run past each row', () => {
  const sessions = [root('r'), fork('a', 'r', '1'), fork('deep', 'a', '2'), fork('b', 'r', '3')]
  const rows = familyRows(sessions[0]!, forksByParent(sessions), never)
  const by = (id: string) => rows.find((row) => row.session.session_id === id)!

  assert.deepEqual(by('r').trunk, [], 'a root has no lines to its left')
  // b is still to come, so a's own column continues past it.
  assert.deepEqual(by('a').trunk, [true])
  assert.equal(by('a').last, false)
  // deep sits under a, and a still has b below it, so the outer column runs
  // through. The last entry is deep's own elbow column, and deep is a's only
  // fork, so nothing continues below it there.
  assert.deepEqual(by('deep').trunk, [true, false])
  assert.equal(by('deep').last, true, "deep is a's only fork, so its elbow closes")
  // b is the youngest, so its elbow closes and nothing continues below it.
  assert.deepEqual(by('b').trunk, [false])
  assert.equal(by('b').last, true)
})

test("an ancestor's column goes blank once its last child has been drawn", () => {
  // r has one fork, a; a has two. The second of a's forks must not draw a line
  // in r's column, because r has nothing left below.
  const sessions = [root('r'), fork('a', 'r', '1'), fork('x', 'a', '2'), fork('y', 'a', '3')]
  const rows = familyRows(sessions[0]!, forksByParent(sessions), never)
  const by = (id: string) => rows.find((row) => row.session.session_id === id)!

  assert.deepEqual(by('a').trunk, [false], 'a is r\'s only fork')
  assert.deepEqual(by('x').trunk, [false, true], 'blank under r, line under a')
  assert.deepEqual(by('y').trunk, [false, false], 'nothing runs past the youngest')
})

test('folding a row hides everything under it and nothing beside it', () => {
  const sessions = [root('r'), fork('a', 'r', '1'), fork('deep', 'a', '2'), fork('b', 'r', '3')]
  const rows = familyRows(sessions[0]!, forksByParent(sessions), (id) => id === 'a')

  assert.deepEqual(
    rows.map((row) => row.session.session_id),
    ['r', 'a', 'b'],
    'the branch under a is hidden; a itself and its sibling stay',
  )
})

test('folding the root leaves only the root', () => {
  const sessions = [root('r'), fork('a', 'r', '1')]
  const rows = familyRows(sessions[0]!, forksByParent(sessions), always)
  assert.deepEqual(rows.map((row) => row.session.session_id), ['r'])
})

test('depth stops growing past the cap, so a long chain cannot walk off the edge', () => {
  const sessions: Session[] = [root('s0')]
  for (let i = 1; i <= MAX_FORK_DEPTH + 3; i += 1) {
    sessions.push(fork(`s${i}`, `s${i - 1}`, String(i)))
  }
  const rows = familyRows(sessions[0]!, forksByParent(sessions), never)
  const deepest = Math.max(...rows.map((row) => row.depth))

  assert.equal(rows.length, sessions.length, 'every session is still drawn')
  assert.equal(deepest, MAX_FORK_DEPTH)
  for (const row of rows) {
    assert.ok(row.trunk.length <= MAX_FORK_DEPTH, 'no row draws more columns than the cap')
  }
})

test('a cycle draws each session once instead of hanging', () => {
  const a = fork('a', 'b', '1')
  const b = fork('b', 'a', '2')
  const rows = familyRows(a, forksByParent([a, b]), never)
  assert.deepEqual(rows.map((row) => row.session.session_id), ['a', 'b'])
})

test('a group node holds roots only, but counts the whole family', () => {
  const sessions = [root('r', '/x/r', HELIOS), fork('a', 'r', '1', '/x/a', HELIOS)]
  const [node] = buildTree(sessions, [HELIOS])

  assert.deepEqual(node?.sessions.map((s) => s.session_id), ['r'], 'the fork is drawn by its parent')
  assert.equal(node?.total, 2, 'a folded family must not shrink the header count')
})

test('grouping by directory keeps a family together even though the fork runs elsewhere', () => {
  // The whole point of the default: a fork gets a worktree of its own, which is
  // a different directory. Grouping by directory must not split the family.
  const sessions = [root('r', '/repo/helios'), fork('a', 'r', '1', '/repo/helios-worktrees/try')]
  const nodes = buildCwdTree(sessions)

  assert.equal(nodes.length, 1, 'one directory node, not two')
  assert.equal(nodes[0]?.key, '/repo/helios', 'the family lives where its root lives')
  assert.deepEqual(nodes[0]?.sessions.map((s) => s.session_id), ['r'])
  assert.equal(nodes[0]?.total, 2)
})

test('an orphan fork gets a node of its own rather than vanishing', () => {
  const nodes = buildCwdTree([fork('lonely', 'gone', '1', '/repo/x')])
  assert.equal(nodes.length, 1)
  assert.deepEqual(nodes[0]?.sessions.map((s) => s.session_id), ['lonely'])
})

File: mobile/test/forks_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:helios/models/session.dart';
import 'package:helios/models/session_group.dart';
import 'package:helios/utils/grouping.dart';

/// The fork tree the phone draws. A port of the desktop's `forks.test.ts`, and
/// pinned for the same reason the grouping tests are: two clients read one
/// daemon, and a family that nests one way here and another way there is two
/// answers to one question.

const helios = SessionGroup(key: 'g_helios', name: 'helios', position: 0);

Session root(String id, {String cwd = '/x/repo', SessionGroup? group}) => Session(
  sessionId: id,
  source: 'claude',
  cwd: cwd,
  project: 'repo',
  status: 'idle',
  createdAt: '2026-09-19T00:00:00Z',
  groupKey: group?.key ?? '',
  groupPath: group == null ? const [] : [group],
);

Session fork(
  String id,
  String parent,
  String at, {
  String cwd = '/x/repo',
  SessionGroup? group,
}) => Session(
  sessionId: id,
  source: 'claude',
  cwd: cwd,
  project: 'repo',
  status: 'idle',
  createdAt: '2026-09-19T00:00:00Z',
  groupKey: group?.key ?? '',
  groupPath: group == null ? const [] : [group],
  forkedFrom: parent,
  forkedAt: at,
);

bool never(String _) => false;
bool always(String _) => true;

void main() {
  test('a fork nests only when its parent is in the same list', () {
    final sessions = [root('a'), fork('b', 'a', '1'), fork('orphan', 'gone', '2')];
    final present = idsOf(sessions);

    expect(isNestedFork(sessions[1], present), isTrue);
    expect(isNestedFork(sessions[2], present), isFalse,
        reason: 'an orphan draws as a root rather than never drawing');
    expect(isNestedFork(sessions[0], present), isFalse);
  });

  test('siblings are ordered by when the branch was taken', () {
    final forks = forksByParent([
      fork('late', 'a', '2026-09-19T12:00:00Z'),
      fork('early', 'a', '2026-09-19T09:00:00Z'),
    ]);
    expect(forks['a']!.map((s) => s.sessionId), ['early', 'late']);
  });

  test('a family is drawn depth-first, parent then its own forks', () {
    final sessions = [
      root('r'),
      fork('a', 'r', '1'),
      fork('deep', 'a', '2'),
      fork('b', 'r', '3'),
    ];
    final rows = familyRows(sessions.first, forksByParent(sessions), never);

    expect(
      rows.map((row) => '${row.session.sessionId}:${row.depth}'),
      ['r:0', 'a:1', 'deep:2', 'b:1'],
    );
  });

  test('the trunk says which ancestor lines run past each row', () {
    final sessions = [
      root('r'),
      fork('a', 'r', '1'),
      fork('deep', 'a', '2'),
      fork('b', 'r', '3'),
    ];
    final rows = familyRows(sessions.first, forksByParent(sessions), never);
    FamilyRow by(String id) =>
        rows.firstWhere((row) => row.session.sessionId == id);

    expect(by('r').trunk, isEmpty, reason: 'a root has no lines to its left');
    // b is still to come, so a's own column carries on past it.
    expect(by('a').trunk, [true]);
    expect(by('a').last, isFalse);
    // deep sits under a, and a still has b below: the outer column runs
    // through. The last entry is deep's own elbow, and it is an only child.
    expect(by('deep').trunk, [true, false]);
    expect(by('deep').last, isTrue);
    expect(by('b').trunk, [false]);
    expect(by('b').last, isTrue);
  });

  test("an ancestor's column goes blank once its last child is drawn", () {
    final sessions = [
      root('r'),
      fork('a', 'r', '1'),
      fork('x', 'a', '2'),
      fork('y', 'a', '3'),
    ];
    final rows = familyRows(sessions.first, forksByParent(sessions), never);
    FamilyRow by(String id) =>
        rows.firstWhere((row) => row.session.sessionId == id);

    expect(by('a').trunk, [false]);
    expect(by('x').trunk, [false, true], reason: 'blank under r, line under a');
    expect(by('y').trunk, [false, false]);
  });

  test('folding a row hides what is under it and nothing beside it', () {
    final sessions = [
      root('r'),
      fork('a', 'r', '1'),
      fork('deep', 'a', '2'),
      fork('b', 'r', '3'),
    ];
    final rows = familyRows(
      sessions.first,
      forksByParent(sessions),
      (id) => id == 'a',
    );
    expect(rows.map((row) => row.session.sessionId), ['r', 'a', 'b']);
  });

  test('folding the root leaves only the root', () {
    final sessions = [root('r'), fork('a', 'r', '1')];
    final rows = familyRows(sessions.first, forksByParent(sessions), always);
    expect(rows.map((row) => row.session.sessionId), ['r']);
  });

  test('depth stops growing past the cap', () {
    final sessions = <Session>[root('s0')];
    for (var i = 1; i <= kMaxForkDepth + 3; i++) {
      sessions.add(fork('s$i', 's${i - 1}', '$i'));
    }
    final rows = familyRows(sessions.first, forksByParent(sessions), never);

    expect(rows.length, sessions.length, reason: 'every session still draws');
    expect(rows.map((row) => row.depth).reduce((a, b) => a > b ? a : b),
        kMaxForkDepth);
    for (final row in rows) {
      expect(row.trunk.length, lessThanOrEqualTo(kMaxForkDepth));
    }
  });

  test('a cycle draws each session once instead of hanging', () {
    final a = fork('a', 'b', '1');
    final b = fork('b', 'a', '2');
    final rows = familyRows(a, forksByParent([a, b]), never);
    expect(rows.map((row) => row.session.sessionId), ['a', 'b']);
  });

  test('a group node holds roots only, but counts the whole family', () {
    final sessions = [
      root('r', group: helios),
      fork('a', 'r', '1', group: helios),
    ];
    final nodes = buildTree(sessions, [helios]);

    expect(nodes.first.sessions.map((s) => s.sessionId), ['r'],
        reason: 'the fork is drawn by its parent');
    expect(nodes.first.total, 2,
        reason: 'a folded family must not shrink the header count');
  });

  test('grouping by directory keeps a family whole', () {
    // The whole point of the default: a fork gets a worktree of its own, which
    // is a different directory. Grouping by directory must not split the pair.
    final sessions = [
      root('r', cwd: '/repo/helios'),
      fork('a', 'r', '1', cwd: '/repo/helios-worktrees/try'),
    ];
    final nodes = buildCwdTree(sessions);

    expect(nodes.length, 1, reason: 'one directory node, not two');
    expect(nodes.first.key, '/repo/helios');
    expect(nodes.first.sessions.map((s) => s.sessionId), ['r']);
    expect(nodes.first.total, 2);
  });

  test('an orphan fork gets a node of its own rather than vanishing', () {
    final nodes = buildCwdTree([fork('lonely', 'gone', '1', cwd: '/repo/x')]);
    expect(nodes.length, 1);
    expect(nodes.first.sessions.map((s) => s.sessionId), ['lonely']);
  });

  test('the fork fields survive a round trip from the daemon', () {
    final session = Session.fromJson({
      'session_id': 'f1',
      'source': 'claude',
      'cwd': '/x',
      'project': 'x',
      'status': 'idle',
      'created_at': '2026-09-19T00:00:00Z',
      'forked_from': 'p1',
      'forked_at': '2026-09-19T10:00:00Z',
      'fork_count': 2,
      'root_session_id': 'p1',
    });

    expect(session.isFork, isTrue);
    expect(session.forkedFrom, 'p1');
    expect(session.forkCount, 2);
    expect(session.rootSessionId, 'p1');
  });
}
